clang/lib/Serialization/ASTWriterDecl.cpp

```diff
@@ -1899,7 +1899,7 @@ void ASTDeclWriter::VisitTemplateTypeParmDecl(TemplateTypeParmDecl *D) {
   Record.push_back(D->wasDeclaredWithTypename());
 
   const TypeConstraint *TC = D->getTypeConstraint();
-  assert((bool)TC == D->hasTypeConstraint());
+  Record.push_back(/*TypeConstraintInitialized=*/TC != nullptr);
   if (TC) {
```

Comment:
What about checking for `D->hasTypeConstraint()` instead of pushing am additional bool?

Reply:
Because `D->hasTypeConstraint()` does not imply that `D->getTypeConstraint()` is non-null, as the code assumed (hence, the assertion).